tasks/password-generator/src/main.rs

```diff
@@ -0,0 +1,172 @@
+// declare our external dependencies
+// for parsing command line arguments
+extern crate clap;
+// for working with random values
+extern crate rand;
+// why use a library for this CLI? Clap lets us quickly build a CLI and
+// lets us focus the below code on password generation
+// not the intricacies building a CLI in rust.
+// Read more about clap here: https://clap.rs/
+// bring the needed structures into scope so we
+// invoke them later in the program
+use clap::{App, Arg};
+use rand::distributions::Alphanumeric;
+use rand::prelude::IteratorRandom;
+use rand::{thread_rng, Rng};
+use std::iter;
+
+// the core logic that creates our password
+fn generate_password(length: usize) -> String {
+    // cache thread_rng for better performance
+    let mut rng = thread_rng();
+    let mut base_password: Vec<char> = iter::repeat(())
+        // the Alphanumeric struct provides 3/4
+        // of the characters for passwords
+        // so we can sample from it
+        .map(|()| rng.sample(Alphanumeric))
+        .take(length)
+        .collect();
+    // create an iterator of required other characters
+    let other_values = "!\"#$%&'()*+,-./:;<=>?@[]^_{|}~";
+    // create a random count of how many other characters to add
+    let mut to_add = rng.gen_range(1, 10);
+    loop {
+        let special = other_values.chars().choose(&mut rng).unwrap();
+        to_add -= 1;
+        base_password[to_add] = special;
+        if to_add == 0 {
+            break;
+        }
+    }
+    // you convert the vector of characters into a string
+    // using the turbofish syntax
+    base_password.iter().collect::<String>()
+}
+// validator function for our first argument
+// picked 15 as the length must be greater than
+// our randomly generated range above (10)
+fn is_valid_length(value: String) -> Result<(), String> {
+    let length = value.parse::<u32>();
+    // if we cannot parse the string into an i32
+    if length.is_err() {
+        Err(String::from("Please provide valid password"))
+    } else if length.unwrap() < 15 {
+        Err(String::from("Mininum password length is 15"))
+    } else {
+        Ok(())
+    }
+}
+fn main() {
+    // create our new CLI
+    // clap provides powerful defaults so we don't have to
+    // write all the logic here
+    // For example, clap takes care of the help and version flags by default
+    let matches = App::new("password-generator")
```

Comment:
I'd use structopt instead of clap.

Reply:
Thanks for the review @euclio! What arguments do you see in favor of using structopt over clap? Or what are some of the benefits?

Reply:
I'm specifically wondering how I can:

1. how I can configure `structopt` to provide custom validation for a given argument?
2. While I can easily change the tool to use named arguments (`-l 50`) I decided to keep it using positional arguments to be as simple as possible. Do you know how to achieve positional arguments using structopt? From briefly reading their documentation it was not clear.
While a bit more verbose I think clap is much easier to work with for rust beginners because of the extensive documentation.

Reply:
Structopt ends up being much less verbose, and is a great demonstration of powerful Rust features, like custom derives. For example, the equivalent structopt for your Clap args is:

```
use structopt::StructOpt;

/// Generate a password according to the Rosetta Code rules: https://rosettacode.org/wiki/Password_generator
#[derive(Debug, StructOpt)]
struct Opt {
    /// Password length.
    #[structopt(default_value = "160")]
    length: u16,

    /// How many passwords to generate.
    #[structopt(default_value = "1")]
    count: i32,
}
```

In structopt, making a non-positional argument is accomplished by not providing a `short` or `long` attribute. For custom validation, it's possible with `validator`, but I don't think it's needed in this case?

Reply:
Ah, that makes a lot of sense for simple programs like this.
Reading the [wiki literally](http://rosettacode.org/wiki/Password_generator) we do not need to enforce length. If you run the program with zero for length it panics because I use the password length to seed the random character range:
```
$ cargo run --release -- 0 
   Compiling password-generator v0.1.0 (rust-rosetta/tasks/password-generator)
    Finished release [optimized] target(s) in 1.49s
     Running `rust-rosetta/target/release/password-generator 0`
thread 'main' panicked at 'UniformSampler::sample_single: low >= high', .cargo/registry/src/github.com-1ecc6299db9ec823/rand-0.7.0/src/distributions/uniform.rs:462:1
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace.
```

Reply:
You could either special case 0 to just print out empty lines, or add a validator to make sure the length is nonzero. I think either would be fine.